File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.js
const filePath = process.platform === 'linux' ? '/dev/stdin' : 'e.txt';
const input = require('fs')
	.readFileSync(filePath)
	.toString()
	.trim()
	.split('\n');
const [n, s] = input.shift().split(' ').map(Number);

const numbers = input[0].split(' ').map(Number);
const sequence = [];
let count = 0;
const visited = new Array(numbers.length).fill(false);

const dfs = (index, max) => {
	if (sequence.length === max) {
		const sum = sequence.reduce((sum, cur) => sum + cur, 0);
		if (sum === s) count++;
	}

	for (let i = index; i < numbers.length; i++) {
		if (!visited[i]) {
			visited[i] = true;
			sequence.push(numbers[i]);
			dfs(i + 1, max);
			sequence.pop();
			visited[i] = false;
		}
	}
};

for (let i = 1; i <= numbers.length; i++) {
	dfs(0, i);
}

console.log(count);

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.py
'''
개수는 20개

n, n*n, n*n*n
'''

from itertools import combinations 

n, target = map(int, input().split())
lst = list(map(int, input().split()))

count = 0
for i in range(1, n + 1):
    rst = list(combinations(lst, i))
    for v in rst:
        if sum(v) == target:
            count += 1

print(count)